File: dataset.py
import os
import re
from pathlib import Path
from typing import List, Dict, Union, Optional
import requests
from bs4 import BeautifulSoup
import torch
from torch.utils.data import Dataset

from tokenizer import ByteLevelTokenizer


class WebScraper:

    HEADERS = {
        "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36",
        "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,text/plain;q=0.8,*/*;q=0.7",
        "Accept-Language": "en-US,en;q=0.9",
    }

    @classmethod
    def scrape_url(cls, url: str, timeout: int = 15) -> Dict[str, str]:
        if not url.startswith(("http://", "https://")):
            url = "https://" + url

        try:
            response = requests.get(url, headers=cls.HEADERS, timeout=timeout, allow_redirects=True)
            response.raise_for_status()

            if response.encoding is None or response.encoding.lower() == "iso-8859-1":
                response.encoding = response.apparent_encoding or "utf-8"

            content_type = response.headers.get("Content-Type", "").lower()
            if "text/plain" in content_type:
                clean_text = response.text.strip()
                title = url.split("/")[-1] or url
                return {
                    "source": url,
                    "title": title,
                    "content": clean_text,
                    "char_count": str(len(clean_text))
                }

            soup = BeautifulSoup(response.text, "html.parser")

            for tag in soup(["script", "style", "nav", "footer", "header", "aside", "noscript", "svg", "form", "iframe"]):
                tag.decompose()

            title_tag = soup.find("title")
            title = title_tag.get_text().strip() if title_tag else url

            paragraphs = []
            for element in soup.find_all(["h1", "h2", "h3", "h4", "p", "li", "blockquote", "pre", "code"]):
                text = element.get_text().strip()
                if text and len(text) > 5:
                    paragraphs.append(text)

            if not paragraphs:
                body = soup.find("body")
                if body:
                    clean_text = body.get_text(separator="\n").strip()
                else:
                    clean_text = soup.get_text(separator="\n").strip()
            else:
                clean_text = "\n\n".join(paragraphs)

            clean_text = re.sub(r"\n{3,}", "\n\n", clean_text)
            clean_text = re.sub(r"[ \t]{2,}", " ", clean_text)

            return {
                "source": url,
                "title": title,
                "content": clean_text,
                "char_count": str(len(clean_text))
            }

        except Exception as e:
            return {
                "source": url,
                "title": url,
                "content": "",
                "error": str(e)
            }


class DocumentReader:

    ENCODINGS = ["utf-8", "utf-8-sig", "latin-1", "cp1252"]

    @classmethod
    def is_binary(cls, file_path: Path) -> bool:
        try:
            with open(file_path, "rb") as f:
                chunk = f.read(1024)
                if b"\x00" in chunk:
                    return True
        except Exception:
            return True
        return False

    @classmethod
    def read_file(cls, file_path: Union[str, Path]) -> Dict[str, str]:
        path = Path(file_path).resolve()
        if not path.exists() or not path.is_file():
            return {"source": str(path), "title": str(path), "content": "", "error": "File not found"}

        if cls.is_binary(path):
            return {"source": str(path), "title": path.name, "content": "", "error": "Binary file skipped"}

        content = ""
        for enc in cls.ENCODINGS:
            try:
                with open(path, "r", encoding=enc, errors="strict") as f:
                    content = f.read()
                break
            except (UnicodeDecodeError, LookupError):
                continue

        if not content:
            try:
                with open(path, "r", encoding="utf-8", errors="replace") as f:
                    content = f.read()
            except Exception as e:
                return {"source": str(path), "title": path.name, "content": "", "error": str(e)}

        clean_text = re.sub(r"\r\n", "\n", content).strip()
        return {
            "source": str(path),
            "title": path.name,
            "content": clean_text,
            "char_count": str(len(clean_text))
        }

    @classmethod
    def read_directory(cls, dir_path: Union[str, Path], recursive: bool = True) -> List[Dict[str, str]]:
        path = Path(dir_path).resolve()
        if not path.exists() or not path.is_dir():
            return []

        supported_exts = {".txt", ".md", ".markdown", ".csv", ".tsv", ".json", ".py", ".js", ".html", ".log", ".rst"}
        documents = []

        glob_func = path.rglob if recursive else path.glob
        for file_path in glob_func("*"):
            if file_path.is_file() and file_path.suffix.lower() in supported_exts:
                doc = cls.read_file(file_path)
                if doc.get("content"):
                    documents.append(doc)

        return documents


class TextChunker:

    @classmethod
    def chunk_text(
        cls,
        text: str,
        tokenizer: ByteLevelTokenizer,
        chunk_size: int = 256,
        overlap: int = 64,
        source_meta: Optional[Dict] = None
    ) -> List[Dict]:
        if not text:
            return []

        source_meta = source_meta or {}
        source = source_meta.get("source", "unknown")
        title = source_meta.get("title", "Untitled")

        paragraphs = text.split("\n\n")
        raw_sections = []
        for p in paragraphs:
            p_clean = p.strip()
            if len(p_clean) > 800:
                sentences = re.split(r"(?<=[.!?])\s+", p_clean)
                buffer = ""
                for s in sentences:
                    if len(buffer) + len(s) < 600:
                        buffer += (" " if buffer else "") + s
                    else:
                        if buffer:
                            raw_sections.append(buffer)
                        buffer = s
                if buffer:
                    raw_sections.append(buffer)
            elif p_clean:
                raw_sections.append(p_clean)

        chunks = []
        chunk_idx = 0
        step = max(1, chunk_size - overlap)

        for sec in raw_sections:
            sec_tokens = tokenizer.encode(sec, add_special_tokens=False)
            if not sec_tokens:
                continue

            if len(sec_tokens) <= chunk_size:
                chunks.append({
                    "chunk_id": f"{source}_{chunk_idx}",
                    "chunk_idx": chunk_idx,
                    "source": source,
                    "title": title,
                    "text": sec,
                    "tokens": sec_tokens
                })
                chunk_idx += 1
            else:
                for i in range(0, len(sec_tokens), step):
                    window = sec_tokens[i:i + chunk_size]
                    if len(window) < 10:
                        continue
                    window_text = tokenizer.decode(window, skip_special_tokens=True)
                    chunks.append({
                        "chunk_id": f"{source}_{chunk_idx}",
                        "chunk_idx": chunk_idx,
                        "source": source,
                        "title": title,
                        "text": window_text,
                        "tokens": window
                    })
                    chunk_idx += 1

        return chunks


class NeuralTextDataset(Dataset):

    def __init__(self, chunks: List[Dict], max_seq_len: int = 256, pad_token_id: int = 0):
        self.samples = []
        self.max_seq_len = max_seq_len
        self.pad_token_id = pad_token_id

        for c in chunks:
            toks = c.get("tokens", [])
            if len(toks) > 1:
                self.samples.append(toks[:max_seq_len])

    def __len__(self) -> int:
        return len(self.samples)

    def __getitem__(self, idx: int) -> Tuple[torch.Tensor, torch.Tensor]:
        tokens = self.samples[idx]
        seq_len = len(tokens)

        if seq_len < self.max_seq_len:
            padded = tokens + [self.pad_token_id] * (self.max_seq_len - seq_len)
            labels = tokens[1:] + [self.pad_token_id] + [-100] * (self.max_seq_len - seq_len)
        else:
            padded = tokens[:self.max_seq_len]
            labels = tokens[1:self.max_seq_len] + [self.pad_token_id]

        input_tensor = torch.tensor(padded, dtype=torch.long)
        label_tensor = torch.tensor(labels, dtype=torch.long)
        return input_tensor, label_tensor

File: src/dataset.rs
use crate::tokenizer::Tokenizer;
use regex::Regex;
use scraper::{Html, Selector};
use std::path::Path;

#[derive(Debug, Clone)]
pub struct RawDocument {
    pub source: String,
    pub title: String,
    pub content: String,
}

pub struct WebScraper;

impl WebScraper {
    pub fn scrape_url(url_str: &str) -> Result<RawDocument, Box<dyn std::error::Error>> {
        let full_url = if !url_str.starts_with("http://") && !url_str.starts_with("https://") {
            format!("https://{}", url_str)
        } else {
            url_str.to_string()
        };

        let client = reqwest::blocking::Client::builder()
            .user_agent("Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36")
            .timeout(std::time::Duration::from_secs(15))
            .build()?;

        let response = client.get(&full_url).send()?;
        let content_type = response
            .headers()
            .get(reqwest::header::CONTENT_TYPE)
            .and_then(|v| v.to_str().ok())
            .unwrap_or("")
            .to_string();

        let body = response.text()?;

        if content_type.contains("text/plain") {
            let title = full_url.split('/').last().unwrap_or(&full_url).to_string();
            return Ok(RawDocument {
                source: full_url,
                title,
                content: body.trim().to_string(),
            });
        }

        let document = Html::parse_document(&body);

        let title_sel = Selector::parse("title").unwrap();
        let title = document
            .select(&title_sel)
            .next()
            .map(|e| e.text().collect::<Vec<_>>().join(" "))
            .unwrap_or_else(|| full_url.clone());

        let p_sel = Selector::parse("h1, h2, h3, h4, p, li, blockquote, pre, code").unwrap();
        let mut paragraphs = Vec::new();

        for element in document.select(&p_sel) {
            let text = element.text().collect::<Vec<_>>().join(" ").trim().to_string();
            if text.len() > 5 {
                paragraphs.push(text);
            }
        }

        let raw_text = if paragraphs.is_empty() {
            let body_sel = Selector::parse("body").unwrap();
            document
                .select(&body_sel)
                .next()
                .map(|e| e.text().collect::<Vec<_>>().join(" "))
                .unwrap_or(body)
        } else {
            paragraphs.join("\n\n")
        };

        let space_re = Regex::new(r"[ \t]{2,}")?;
        let newline_re = Regex::new(r"\n{3,}")?;

        let cleaned = space_re.replace_all(&raw_text, " ");
        let cleaned = newline_re.replace_all(&cleaned, "\n\n");

        Ok(RawDocument {
            source: full_url,
            title: title.trim().to_string(),
            content: cleaned.trim().to_string(),
        })
    }
}

pub struct DocumentReader;

impl DocumentReader {
    pub fn read_file<P: AsRef<Path>>(path: P) -> Result<RawDocument, Box<dyn std::error::Error>> {
        let path = path.as_ref();
        let content = std::fs::read_to_string(path)?;
        let title = path
            .file_name()
            .and_then(|n| n.to_str())
            .unwrap_or("Untitled")
            .to_string();

        Ok(RawDocument {
            source: path.to_string_lossy().to_string(),
            title,
            content: content.trim().to_string(),
        })
    }

    pub fn read_directory<P: AsRef<Path>>(dir: P, recursive: bool) -> Vec<RawDocument> {
        let mut docs = Vec::new();
        let supported_exts = ["txt", "md", "csv", "json", "log", "rst", "py", "rs", "js", "html"];

        let walk_dir = |p: &Path, docs: &mut Vec<RawDocument>, recurse: bool| {
            if let Ok(entries) = std::fs::read_dir(p) {
                for entry in entries.flatten() {
                    let entry_path = entry.path();
                    if entry_path.is_file() {
                        if let Some(ext) = entry_path.extension().and_then(|e| e.to_str()) {
                            if supported_exts.contains(&ext.to_lowercase().as_str()) {
                                if let Ok(doc) = DocumentReader::read_file(&entry_path) {
                                    if !doc.content.is_empty() {
                                        docs.push(doc);
                                    }
                                }
                            }
                        }
                    } else if entry_path.is_dir() && recurse {
                        let name = entry_path.file_name().and_then(|n| n.to_str()).unwrap_or("");
                        if !name.starts_with('.') && name != "target" && name != "node_modules" {
                            let _ = DocumentReader::read_directory(&entry_path, true);
                        }
                    }
                }
            }
        };

        walk_dir(dir.as_ref(), &mut docs, recursive);
        docs
    }
}

pub struct TextChunker;

impl TextChunker {
    pub fn chunk_text(
        text: &str,
        tokenizer: &Tokenizer,
        chunk_size: usize,
        _overlap: usize,
    ) -> Vec<(String, Vec<usize>)> {
        let mut chunks = Vec::new();
        let paragraphs: Vec<&str> = text.split("\n\n").map(|p| p.trim()).filter(|p| !p.is_empty()).collect();

        let mut current_text = String::new();
        let mut current_tokens = Vec::new();

        for para in paragraphs {
            let para_tokens = tokenizer.encode(para, false);

            if current_tokens.len() + para_tokens.len() <= chunk_size {
                if !current_text.is_empty() {
                    current_text.push_str("\n\n");
                }
                current_text.push_str(para);
                current_tokens.extend(para_tokens);
            } else {
                if !current_tokens.is_empty() {
                    chunks.push((current_text.clone(), current_tokens.clone()));
                }

                current_text = para.to_string();
                current_tokens = para_tokens;
            }
        }

        if !current_tokens.is_empty() {
            chunks.push((current_text, current_tokens));
        }

        chunks
    }
}

